readme.md
# lrio
[![js-standard-style](https://img.shields.io/badge/code%20style-standard-brightgreen.svg?style=flat)](https://github.com/feross/standard)
[![travis][travis-image]][travis-url]
[![npm][npm-image]][npm-url]
[![downloads][downloads-image]][downloads-url]

[travis-image]: https://img.shields.io/travis/nodys/lrio.svg?style=flat&branch=master
[travis-url]: https://travis-ci.org/nodys/lrio
[npm-image]: https://img.shields.io/npm/v/lrio.svg?style=flat
[npm-url]: https://npmjs.org/package/lrio
[downloads-image]: https://img.shields.io/npm/dm/lrio.svg?style=flat
[downloads-url]: https://npmjs.org/package/lrio

> Minimalist websocket broadcast server and browser client

## Feature

- Broadcast messages using a dedicated channel
- Attach as many websocket server you want without collision
- Client for browserify
- Based on [WebSocket-Node](https://github.com/Worlize/WebSocket-Node)

## Usage

```bash
npm install --save lrio
```

### Server side

```javascript

var http   = require('http');
var lrio   = require('lrio');
var server = http.createServer();

var A = lrio(server, 'myChannelA');
var B = lrio(server, 'myChannelB');

server.listen(8080);

// Broadcast message on myChannelA
A.broadcast({my:42})

// Broadcast message on myChannelB
B.broadcast({foo:'bar'})

// Other events :
A.on('join',  function(client) {})
A.on('leave', function(client) {})

```
Notice that `client` are [WebSocketConnection](https://github.com/Worlize/WebSocket-Node/wiki/Documentation#websocketconnection) instances

### Client side (using browserify)

```javascript

var lrio   = require('lrio');

var A = lrio('myChannelA');
var B = lrio('myChannelB');

A.on('message', function(message) { console.log(message) })
B.on('message', function(message) { console.log(message) })

// Other events:
A.on('open', function() {})
A.on('close', function() {})
A.on('error', function() {})

A.close(); // Close websocket
A.connect(); // Re-connect  (only if closed)

```

## License

[The MIT license](./LICENSE) • Novadiscovery


[![js-standard-style](https://cdn.rawgit.com/feross/standard/master/badge.svg)](https://github.com/feross/standard)


changelog.md
Changelog
=========

v1.2.1 - Proud Frog (2015-05-29) 
----------------------------------------------------------------------

  - test: Use iojs for test


v1.2.0 - Generous Snake (2015-05-29) 
----------------------------------------------------------------------

  - doc: Update badges
  - test: Upgrade test according to dependency api changes


v1.1.0 - Fluffy Owl (2015-01-22) 
----------------------------------------------------------------------

  - feat: No more HEAD request to ping server and introduce timeout


v1.0.0 - Fast Snake (2014-10-22) 
----------------------------------------------------------------------

  - refactor: Move to another repository
  - test: Update test coverage


v0.0.3 - Discreet Lion (2014-10-02) 
----------------------------------------------------------------------

  - refactor: Use emit() in place of trigger() (now an alias)


v0.0.2 - Calamitous Cheetah (2014-10-01) 
----------------------------------------------------------------------

  - doc: Fix travis link


v0.0.1 - Lovable Horse (2014-10-01) 
----------------------------------------------------------------------

  - doc: Update documentation
  - test: Full test coverage
  - test: Add full server tests




test/server.test.js
/* global describe it */
var expect = require('expect.js')
var lrio = (process.env.COVERAGE ? require('../src-cov/server.js') : require('../src/server.js'))
var EventEmitter = require('events').EventEmitter

describe('lrio', function () {
  describe('server', function () {
    it('should be connected to an http(s) server', function () {
      var listeners = {}
      var mockServer = {
        on: function (type, listener) {
          listeners[type] = listener
        }
      }
      var lrServer = lrio(mockServer)
      expect(lrServer).to.have.keys('broadcast')
      expect(lrServer).to.have.keys('wsServer')
      expect(lrServer).to.have.keys('getClients')
      expect(listeners).to.have.keys('upgrade')
    })

    it('should accept client and broadcast messages', function () {
      var mockServer = new EventEmitter()
      var lrServer = lrio(mockServer)
      var rejected = false
      var accepted = false
      var protocol, origin, closeListener, message

      var mockRequest = {
        requestedProtocols: ['lrio-protocol-default'],
        origin: '--origin--',
        reject: function () {
          rejected = true
        },
        accept: function (_protocol, _origin) {
          protocol = _protocol
          origin = _origin
          accepted = true

          // Return mock connection
          return mockConnection
        }
      }

      var mockConnection = {
        remoteAddress: '--remote--',
        sendUTF: function (_message) {
          message = _message
        },
        on: function (type, listener) {
          if (type === 'close') {
            closeListener = listener
          }
        }
      }

      lrServer.wsServer.emit('request', mockRequest)

      expect(rejected).to.be(false)
      expect(accepted).to.be(true)
      expect(protocol).to.eql('lrio-protocol-default')
      expect(origin).to.eql('--origin--')

      // Expect client to contain the mocked client ():
      expect(lrServer.getClients()).to.contain(mockConnection)

      // Broadcast a change:
      lrServer.broadcast({type: 'change', uid: 'myUid', src: 'mySrc'})
      expect(message).to.be.eql(JSON.stringify({type: 'change', uid: 'myUid', src: 'mySrc'}))

      // Client close connection:
      closeListener()
      expect(lrServer.getClients()).not.to.contain(mockConnection)

    })

    it('should ignore from other protocols', function () {
      var mockServer = new EventEmitter()
      var lrServer = lrio(mockServer)
      var rejected = false
      var accepted = false

      var mockRequest = {
        requestedProtocols: ['other-protocol'],
        reject: function () {
          rejected = true
        },
        accept: function (_protocol, _origin) {
          console.log('Accepted')
          accepted = true
        }
      }

      lrServer.wsServer.emit('request', mockRequest)

      expect(rejected).to.be(false)
      expect(accepted).to.be(false)

    })

    it('should reject invalid client', function () {
      var mockServer = new EventEmitter()

      function validateClient (client) {
        return false
      }

      var lrServer = lrio(mockServer, null, validateClient)
      var rejected = false
      var accepted = false

      var mockRequest = {
        requestedProtocols: ['lrio-protocol-default'],
        reject: function () {
          rejected = true
        },
        accept: function (_protocol, _origin) {
          accepted = true
        }
      }

      lrServer.wsServer.emit('request', mockRequest)

      expect(rejected).to.be(true)
      expect(accepted).to.be(false)

    })

  })

})


test/browser.test.js
/* global describe it before */
var expect = require('expect.js')

var jsdom = require('jsdom')
var browserify = require('browserify')
var concatStream = require('concat-stream')
var join = require('path').join

describe('lrio', function () {
  describe('browser', function () {
    var bBundle

    before(function (done) {
      var b = browserify()
      var lrio = process.env.COVERAGE ? join(__dirname, '../src-cov/browser.js') : join(__dirname, '../src/browser.js')
      b.require(lrio, {
        expose: 'lrio'
      })
      b.bundle().pipe(concatStream(function (result) {
        bBundle = result.toString()
        done()
      }))
    })

    function readJsdomError (errors) {
      var msg = errors.reduce(function (memo, item) {
        memo += item.type + ' ' + item.message + ':\n  ' + item.data.error.message
        return memo
      }, '')
      return new Error(msg)
    }

    function mockXMLHttpRequest (forChannel) {
      return function () {
        var self = this
        self.readyState = 2
        self.getResponseHeader = function (header) {
          if (header === 'X-Lrio-' + forChannel) return 'enabled'
        }
        self.open = function () {}
        self.send = function () {
          self.onreadystatechange()
        }
      }
    }

    function mockWebSocket () {
      function mock (remote, protocol) {
        mock.sockets[protocol] = this
      }
      mock.sockets = {}
      return mock
    }

    function watchClient (client) {
      var result = {
        eventCalled: [],
        message: null,
        unwatch: function () {
          client.off('message', onmessage)
          client.off('error', onerror)
          client.off('open', onopen)
          client.off('close', onclose)
        }
      }
      function onmessage (_message) {
        result.eventCalled.push('message')
        result.message = _message
      }
      function onerror () {
        result.eventCalled.push('error')
      }
      function onopen () {
        result.eventCalled.push('open')
      }
      function onclose () {
        result.eventCalled.push('close')
      }

      client.on('message', onmessage)
      client.on('error', onerror)
      client.on('open', onopen)
      client.on('close', onclose)
      return result
    }

    it('should listen for server messages on the given channel', function (done) {
      var channel = 'myChannel'
      jsdom.env({
        'html': '<html></html>',
        'src': [bBundle],
        'created': function (errors, window) {
          if (errors && errors.length) {
            return done(readJsdomError(errors))
          }
          window.XMLHttpRequest = mockXMLHttpRequest('myChannel')
          window.WebSocket = mockWebSocket()
          // console.log('jscov ?', global['_$jscoverage'])
          if (typeof (global['_$jscoverage']) !== 'undefined') {
            window['_$jscoverage'] = global['_$jscoverage']
          }
        },
        'done': function (errors, window) {
          if (errors && errors.length) {
            return done(readJsdomError(errors))
          }

          var client1 = window.require('lrio')(channel)
          var watch1 = watchClient(client1)

          var client2 = window.require('lrio')('anotherChannel')
          var watch2 = watchClient(client2)

          var socket1 = window.WebSocket.sockets['lrio-protocol-myChannel']
          var socket2 = window.WebSocket.sockets['lrio-protocol-anotherChannel']

          socket1.onopen()
          socket1.onmessage({
            data: JSON.stringify({
              myData: '42'
            })
          })
          socket1.onclose()
          try {
            socket1.onerror() // Mocha now detect an error...
          } catch(e) {}

          // Test client.off(event, listener) : see watchClient()
          watch1.unwatch()
          socket1.onmessage({
            data: JSON.stringify({
              myData: '42'
            })
          })

          expect(watch1.message).to.eql({
            myData: '42'
          })
          expect(watch1.eventCalled).to.eql([
            'open',
            'message',
            'close',
            'error'
          ])

          // Socket2 (anotherChannel) must not even be created
          // (the mockXMLHttpRequest accept only myChannel)
          expect(socket2).not.to.eql(undefined)
          expect(watch2.message).to.eql(null)
          expect(watch2.eventCalled).to.eql([])

          done()
        }
      })
    })

  })

})


src/browser.js
module.exports = function (channel, options) {
  channel = channel || 'default'
  var protocol = 'lrio-protocol-' + channel

  var client = {}
  var listeners = {}

  // States
  var OFF = 0
  var WAITING = 1
  var ESTABLISHED = 2
  var state = OFF

  // Options:
  options = options || {}

  // Auto connect:
  if (options.auto === false) {
    options.auto = false
  } else {
    options.auto = true
  }
  options.timeout = typeof (options.timeout) === 'undefined' ? 3000 : options.timeout

  // Event emitter
  client.emit = client.trigger = function (type, data) {
    listeners[type] = listeners[type] || []
    listeners[type].forEach(function (listener) {
      listener.call(null, data)
    })
  }

  client.on = function (type, listener) {
    listeners[type] = listeners[type] || []
    listeners[type].push(listener)
  }

  client.off = client.removeListener = function (type, listener) {
    if (!listeners[type]) return
    listeners[type] = listeners[type].filter(function (_listener) {
      return listener !== _listener
    })
  }

  client.socket = null

  client.close = function () {
    client.socket.close()
  }

  client.connect = function (callback) {
    if (state !== OFF) {
      return
    }
    state = WAITING

    callback = callback || function () {}
    var loc = document.location
    var remote = (loc.protocol === 'https:' ? 'wss:' : 'ws:') + '//' + loc.host + '/'

    try {
      client.socket = new window.WebSocket(remote, protocol)
    } catch(e) {
      return
    }

    if (options.timeout) {
      setTimeout(function () {
        if (state === WAITING) {
          try {
            client.close()
            client.emit('timeout')
          } catch(e) {}
        }
      }, options.timeout)
    }

    client.socket.onmessage = function (event) {
      var message = JSON.parse(event.data)
      client.trigger('message', message)
    }
    client.socket.onerror = function (error) {
      client.trigger('error', error)
      client.close()
    }
    client.socket.onopen = function (event) {
      state = ESTABLISHED
      client.trigger('open', event)
      callback()
    }
    client.socket.onclose = function (event) {
      client.trigger('close', event)
      state = OFF
      client.socket = null
    }
  }

  if (options.auto) client.connect()

  return client

}


src/server.js
var Debug = require('debug')
var WebSocketServer = require('websocket').server
var EventEmitter = require('events').EventEmitter

/**
 * Attach a lrio server to a http(s) server
 *
 * Exemple:
 *
 *    var lrio       = require('lrio')
 *    var http       = require('http')
 *    var server     = http.createServer(myServerApplication)
 *    var myChannel  = lrio(server,'myChannel')
 *    server.listen(8080)
 *
 *    myChannel.broadcast({data:42})
 *
 *    myChannel.getClients().forEach(function(client) { ... })
 *
 * @param  {http.Server|https.Server} server
 *         A node http(s) server instance
 *
 * @param  {String} [channel]
 *         Optional channel name (used as websocket protocol
 *         and server validation)
 *
 * @param  {Function} [validateClient]
 *         Optional websocket request validator, request is rejected
 *         if the returned value is false
 *
 * @return {Object}
 *         An EventEmitter object with:
 *
 *         - `broadcast` {Function}: Broadcast something to all active clients
 *         - `getClients` {Function}: Get all active clients
 *         - `wsServer` {WebSocketServer}: The WebSocketServer instance
 *
 *         Events:
 *
 *         - `join` : When a client is accepted (argument: client's  WebSocketConnection instance)
 *         - `leave` : When a client is closed (argument: client's WebSocketConnection instance)
 *
 *         Other event may be attached to the WebSocketServer instance or the
 *         WebSocketConnection instances (see https://github.com/Worlize/WebSocket-Node)
 */
module.exports = function (server, channel, validateClient) {
  var ctrl = new EventEmitter()
  ctrl.channel = channel || 'default'
  ctrl.protocol = 'lrio-protocol-' + ctrl.channel
  ctrl.wsServer = new WebSocketServer({httpServer: server, autoAcceptConnections: false })
  ctrl.clients = []
  var debug = Debug('lrio:' + ctrl.channel)

  validateClient = validateClient || function () { return true }

  debug('connect to server on channel %s', channel)

  ctrl.wsServer.on('request', function (request) {
    if (!~request.requestedProtocols.indexOf(ctrl.protocol)) {
      return
    }

    if (!validateClient(request)) {
      return request.reject()
    }

    var connection = request.accept(ctrl.protocol, request.origin)
    ctrl.clients.push(connection)
    debug('ws: Peer connected %s', connection.remoteAddress)
    ctrl.emit('join', connection)

    connection.on('close', function (reasonCode, description) {
      ctrl.clients = ctrl.clients.filter(function (client) {
        return client !== connection
      })
      debug('ws: Peer disconnected %s', connection.remoteAddress)
      ctrl.emit('leave', connection)
    })
  })

  ctrl.broadcast = function (data) {
    var message = JSON.stringify(data)
    ctrl.clients.forEach(function (client) {
      client.sendUTF(message)
    })
  }

  ctrl.getClients = function () {
    return ctrl.clients
  }

  return ctrl
}
